### PR TITLE
Improve viewport agg performance

### PR DESCRIPTION
**Pull request by davidmanzanares**
Fix https://github.com/CartoDB/carto-vl/issues/377

Improve performance on viewport aggregations with points in 1.9x

Optimizations:
- Remove the call to `_featureFromDataFrame` by caching its results. Tradeoff memory for CPU time.
- Remove the call to `getAspect()` in the _isFeatureInViewport since it is a expensive call in a loop
- `Little Things`

Previously on CARTO VL:
vperf-prev

On the next episode of CARTO VL:
vperf-fix

### CODE REVIEW

File: src/core/renderer.js [hunk 1]
```diff
@@ -165,11 +165,35 @@ class Renderer {
 
         viewportExpressions.forEach(expr => expr._resetViewportAgg());
 
+        const viewportExpressionsLength = viewportExpressions.length;
+
         // Avoid acumulating the same feature multiple times keeping a set of processed features (same feature can belong to multiple dataframes).
         const processedFeaturesIDs = new Set();
 
+
+        this.aspect= this.gl.canvas.width / this.gl.canvas.height;
```
Comment (IagoLast) [inline src/core/renderer.js]: instead mutating this, I`d create a constant and i´d use this constant as a parameter in `_isFeatureInViewport``

```js
constant ASPECT_RATIO = this.gl.canvas.width / this.gl.canvas.height;

// ...

 if (!this._isFeatureInViewport(dataframe, i, ASPECT_RATIO)) {
// ...

File: src/core/renderer.js [hunk 2]
```diff
@@ -179,24 +203,25 @@ class Renderer {
...
 
-                const feature = this._featureFromDataFrame(dataframe, i);
-
                 // Ignore filtered features
                 if (viz.filter.eval(feature) < FILTERING_THRESHOLD) {
                     continue;
                 }
 
-                viewportExpressions.forEach(viewportExpression => viewportExpression.accumViewportAgg(feature));
+                for (let j = 0; j < viewportExpressionsLength; j++) {
```
Comment (IagoLast) [inline src/core/renderer.js]: what was the x% improvement of using a for loop?

Comment (davidmanzanares) [reply src/core/renderer.js]: I didn't check it with just that change, sorry, anyway, the for loop is quite hot

File: src/core/renderer.js [hunk 3]
```diff
@@ -165,11 +165,35 @@ class Renderer {
...
         const processedFeaturesIDs = new Set();
 
+
+        this.aspect= this.gl.canvas.width / this.gl.canvas.height;
         dataframes.forEach(dataframe => {
+            const propertyNames = Object.keys(dataframe.properties);
+            const propertyNamesLength = propertyNames.length;
+
             for (let i = 0; i < dataframe.numFeatures; i++) {
+                let feature;
```
Comment (jgoizueta) [inline src/core/renderer.js]: Shouldn't all this be encapsulated in a Dataframe method?

Comment (IagoLast) [reply src/core/renderer.js]: In fact this logic were in the `_featureFromDataFrame` method

**Comment by davidmanzanares** [comment]
:+1: Done, pls, re CR

File: src/renderer/Renderer.js [hunk 4]
```diff
@@ -184,17 +188,20 @@ export default class Renderer {
...
+                }
             }
         });
     }
 
     /**
      * Check if the feature at the "index" position of the given dataframe is in the renderer viewport.
+     * NOTE: requires `this.aspect` to be set
      */
     _isFeatureInViewport(dataframe, index) {
```
Comment (IagoLast) [inline src/renderer/Renderer.js]: what about passing the aspect as a parameter?

Comment (davidmanzanares) [reply src/renderer/Renderer.js]: ok...

File: src/renderer/Renderer.js [hunk 5]
```diff
@@ -219,11 +226,20 @@ export default class Renderer {
...
-        const propertyNames = Object.keys(dataframe.properties);
-        const feature = {};
-        for (let i = 0; i < propertyNames.length; i++) {
-            const name = propertyNames[i];
-            feature[name] = dataframe.properties[name][index];
+        let feature;
+        if (!dataframe.cachedFeatures){
+            dataframe.cachedFeatures = [];
+        }
+        if (!dataframe.cachedFeatures[index]) {
```
Comment (IagoLast) [inline src/renderer/Renderer.js]: Little refactor 🌈 
```js
if (dataframe.cachedFeatures[index]) {
    return dataframe.cachedFeatures[index];
}

const feature = {};
const propertyNames = Object.keys(dataframe.properties);
for (let i = 0; i < propertyNames.length; i++) {
    const name = propertyNames[i];
    feature[name] = dataframe.properties[name][index];
}
dataframe.cachedFeatures.push(feature);
return feature;
```

**Comment by IagoLast** [comment]
> That is not correct since the dataframe.cachedFeatures is uninitialized the first time.
 
are you sure?

if the feature is cached --> return feature from cache
otherwise --> generate feature, save it on cache and return it


I mean... the full method would be like:


```js

if (!dataframe.cachedFeatures){
    dataframe.cachedFeatures = [];
}

if (dataframe.cachedFeatures[index]) {
    return dataframe.cachedFeatures[index];
}

const feature = {};
const propertyNames = Object.keys(dataframe.properties);
for (let i = 0; i < propertyNames.length; i++) {
    const name = propertyNames[i];
    feature[name] = dataframe.properties[name][index];
}
dataframe.cachedFeatures.push(feature);
return feature;
```

**Comment by davidmanzanares** [comment]
Ah, that's another code snippet xD

File: debug/viewportagg-perf.html [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+      apiKey: 'default_public'
+    });
+
+    // Define layer
+    const source = new carto.source.Dataset('taxi_50k');
+    const viz = new carto.Viz(`
+      width: ($fare_amount/15000)*zoom()
+      filter: animation($tpep_pickup_datetime,10,fade(0.2,0.2))
+      color: rgba(255,255,255,0.2)
+      strokeColor: ramp(linear($tpep_pickup_datetime,time('2016-05-01T00:08:39Z'),time('2016-05-06T10:29:11Z')),[#CF1C90,#F8CA00,#9BBD4D,#07A5AB,#B100EB])
```
Comment (IagoLast) [inline debug/viewportagg-perf.html]: maybe some space here :trollface:

Comment (davidmanzanares) [reply debug/viewportagg-perf.html]: That's a debug example

**Pull request merged**